ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/V2/LSM.hs

```diff
@@ -0,0 +1,478 @@
+import qualified Data.Vector.Primitive as VP
+import Data.Void
+import Database.LSMTree (Session, Table)
+import qualified Database.LSMTree as LSM
+import NoThunks.Class
+import Ouroboros.Consensus.Block
+import Ouroboros.Consensus.Config
+import Ouroboros.Consensus.Ledger.Abstract
+import Ouroboros.Consensus.Ledger.Extended
+import Ouroboros.Consensus.Ledger.SupportsProtocol
+import qualified Ouroboros.Consensus.Ledger.Tables.Diff as Diff
+import Ouroboros.Consensus.Ledger.Tables.Utils
+import Ouroboros.Consensus.Storage.LedgerDB.API
+import Ouroboros.Consensus.Storage.LedgerDB.Args
+import Ouroboros.Consensus.Storage.LedgerDB.Snapshots
+import Ouroboros.Consensus.Storage.LedgerDB.TraceEvent
+import qualified Ouroboros.Consensus.Storage.LedgerDB.V2.Args as V2
+import Ouroboros.Consensus.Storage.LedgerDB.V2.LedgerSeq
+import Ouroboros.Consensus.Util
+import Ouroboros.Consensus.Util.Args
+import Ouroboros.Consensus.Util.CRC
+import Ouroboros.Consensus.Util.Enclose
+import Ouroboros.Consensus.Util.IOLike
+import Ouroboros.Consensus.Util.IndexedMemPack
+import System.FS.API
+import qualified System.FS.BlockIO.API as BIO
+import System.FS.BlockIO.IO
+import System.Random
+import Prelude hiding (read)
+
+-- | Type alias for convenience
+type UTxOTable m = Table m TxInBytes TxOutBytes Void
+
+instance NoThunks (Table m txin txout Void) where
+  showTypeOf _ = "Table"
+  wNoThunks _ _ = pure Nothing
+
+data LSMClosedExn = LSMClosedExn
+  deriving (Show, Exception)
+
+{-------------------------------------------------------------------------------
+  TxOuts
+-------------------------------------------------------------------------------}
+
+newtype TxOutBytes = TxOutBytes {unTxOutBytes :: LSM.RawBytes}
+
+toTxOutBytes :: IndexedMemPack (l EmptyMK) (TxOut l) => l EmptyMK -> TxOut l -> TxOutBytes
+toTxOutBytes st txout =
+  let barr = indexedPackByteArray True st txout
+   in TxOutBytes $ LSM.RawBytes (VP.Vector 0 (PBA.sizeofByteArray barr) barr)
+
+fromTxOutBytes :: IndexedMemPack (l EmptyMK) (TxOut l) => l EmptyMK -> TxOutBytes -> TxOut l
+fromTxOutBytes st (TxOutBytes (LSM.RawBytes (VP.Vector _ _ barr))) =
+  indexedUnpackError st barr
+
+instance LSM.SerialiseValue TxOutBytes where
+  serialiseValue = unTxOutBytes
+  deserialiseValue = TxOutBytes
+
+deriving via LSM.ResolveAsFirst TxOutBytes instance LSM.ResolveValue TxOutBytes
+
+{-------------------------------------------------------------------------------
+  TxIns
+-------------------------------------------------------------------------------}
+
+newtype TxInBytes = TxInBytes {unTxInBytes :: LSM.RawBytes}
+
+toTxInBytes :: MemPack (TxIn l) => Proxy l -> TxIn l -> TxInBytes
+toTxInBytes _ txin =
+  let barr = packByteArray True txin
+   in TxInBytes $ LSM.RawBytes (VP.Vector 0 (PBA.sizeofByteArray barr) barr)
+
+fromTxInBytes :: MemPack (TxIn l) => Proxy l -> TxInBytes -> TxIn l
+fromTxInBytes _ (TxInBytes (LSM.RawBytes (VP.Vector _ _ barr))) =
+  unpackError barr
```

Comment:
I guess it is fine to throw an imprecise exception here. We will likely enter a crash restart loop, so we might want to have special logging for this case. Maybe document this and potentially wrap the exception?